### PR TITLE
Fix checkMatrix usage in some tests

### PR DESCRIPTION
**Pull request by camelto2**
Please review the [developer documentation](https://github.com/QMCPACK/qmcpack/wiki/Development-workflow)
on the wiki of this project that contains help and requirements.

## Proposed changes

While trying to get a new feature tested that I'm working on, I noticed that a few unit tests have incorrect usage of the "checkMatrix". Those tests were just calling checkMatrix(matA, matB) and not getting the result and checking to see if the matrices agree or not. 

This means that some of those features could actually be failing and we have been missing it since the checks weren't correct. I think I caught all instances here.

## What type(s) of changes does this code introduce?
_Delete the items that do not apply_

- Bugfix
- Testing changes (e.g. new unit/integration/performance tests)

### Does this introduce a breaking change?

- No

## What systems has this change been tested on?
macOS

## Checklist

_Update the following with a yes where the items apply. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- Yes. This PR is up to date with current the current state of 'develop'
- Yes. Code added or changed in the PR has been clang-formatted
- Yes. This PR adds tests to cover any new code, or to catch a bug that is being fixed _(Kind of, it is fixing existing tests)_
- No. Documentation has been added (if appropriate)

### CODE REVIEW

**Comment by camelto2** [comment]
Note that there are failing tests only from the files where I fixed the checkMatrix call to actually see if the matrices agree or not. Seems like there could be lingering bugs

**Comment by ye-luo** [comment]
Test this please

**Comment by ye-luo** [comment]
@camelto2 remaining failure are from the added code?

**Comment by camelto2** [comment]
Looks like it would still be tolerance actually. I think to add a tolerance argument into checkMatrix and that would solve the mixed precision issue

**Comment by camelto2** [comment]
Ok, so the failures were due to the fact that the checkMatrix function uses Catch2 Approx, and was always using the default epsilon. For mixed precision applyRotation in the splines, the laplacian was failing tests due to that default tolerance. 

I added some functionality which allows us to pass in a desired tolerance for the element by element comparison in checkMatrix. Also added some testing of that inside the unit test for checkMatrix.

File: src/Utilities/for_testing/checkMatrix.hpp [hunk 1]
```diff
@@ -50,9 +51,13 @@ struct CheckMatrixResult
...
  *  \param[in] b_mat     - the matrix to check
  *  \param[in] check_all - if true continue to check matrix elements after failure
+ *  \param[in] eps       - add a tolerance for Catch Approx checks. Default to same as in Approx. 
  */
 template<class M1, class M2>
-CheckMatrixResult checkMatrix(M1& a_mat, M2& b_mat, const bool check_all = false)
+CheckMatrixResult checkMatrix(M1& a_mat,
+                              M2& b_mat,
+                              const bool check_all              = false,
+                              const double eps = std::numeric_limits<float>::epsilon() * 100)
```
Comment (ye-luo) [inline src/Utilities/for_testing/checkMatrix.hpp:60]: I don't feel right with this hidden choice. Could you use std::optional and default to `std::nullopt_t`

Comment (camelto2) [reply src/Utilities/for_testing/checkMatrix.hpp]: Think I have this working with optional.

**Review by ye-luo — APPROVED** [review]

**Comment by ye-luo** [comment]
Test this please

**Pull request merged**